flight/PiOS/Common/pios_port.c

```diff
@@ -0,0 +1,623 @@
+                PIOS_Assert(0);
+        }
+        
+        uintptr_t pios_hsum_id;
+        if (PIOS_HSUM_Init(&pios_hsum_id, pios_usart_com_driver,
+                           pios_usart_hsum_id, *proto)) {
+                PIOS_Assert(0);
+        }
+        
+        uintptr_t pios_hsum_rcvr_id;
+        if (PIOS_RCVR_Init(&pios_hsum_rcvr_id, &pios_hsum_rcvr_driver, pios_hsum_id)) {
+                PIOS_Assert(0);
+        }
+        pios_rcvr_group_map[channelgroup] = pios_hsum_rcvr_id;
+}
+#endif
+
+void PIOS_HAL_configure_port(HwSharedPortTypesOptions port_type,
+	const struct pios_usart_cfg *usart_port_cfg, 
+	const struct pios_com_driver *com_driver, 
+	uint32_t *i2c_id,
+	const struct pios_i2c_adapter_cfg *i2c_cfg,
+	const struct pios_ppm_cfg *ppm_cfg,
+	uint32_t led_id,
+/* TODO: future work to factor most of these away */
+	const struct pios_usart_cfg *usart_dsm_hsum_cfg,
+	const struct pios_dsm_cfg *dsm_cfg,
+	int dsm_bind,
+	const struct pios_usart_cfg *sbus_rcvr_cfg,
+	const struct pios_sbus_cfg *sbus_cfg,
+	bool sbus_toggle
+	)
+{
+	uintptr_t port_driver_id;
+	uintptr_t *target=NULL, *target2=NULL;;
+
+        switch (port_type) {
+		case HWSHARED_PORTTYPES_I2C:
+#if defined(PIOS_INCLUDE_I2C)
+			if (i2c_id && i2c_cfg) {
+				if (PIOS_I2C_Init(i2c_id, i2c_cfg)) {
+					PIOS_Assert(0);
+				}
+				if (PIOS_I2C_CheckClear(*i2c_id) != 0)
+					PIOS_HAL_panic(led_id, 6);
+			}
+#endif  /* PIOS_INCLUDE_I2C */
+			break;
+		case HWSHARED_PORTTYPES_PPM:
+#if defined(PIOS_INCLUDE_PPM)
+			if (ppm_cfg) { 
+				uintptr_t pios_ppm_id;
+				PIOS_PPM_Init(&pios_ppm_id, ppm_cfg);
+
+				uintptr_t pios_ppm_rcvr_id;
+				if (PIOS_RCVR_Init(&pios_ppm_rcvr_id, &pios_ppm_rcvr_driver, pios_ppm_id)) {
+					PIOS_Assert(0);
+				}
+				pios_rcvr_group_map[MANUALCONTROLSETTINGS_CHANNELGROUPS_PPM] = pios_ppm_rcvr_id;
+			}
+#endif  /* PIOS_INCLUDE_PPM */
+			break;
+
+		case HWSHARED_PORTTYPES_DISABLED:
+			break;
+		case HWSHARED_PORTTYPES_TELEMETRY:
+			PIOS_HAL_configure_com(usart_port_cfg, PIOS_COM_TELEM_RF_RX_BUF_LEN, PIOS_COM_TELEM_RF_TX_BUF_LEN, com_driver, &port_driver_id);
+			// Name of this should change XXX
+			target = &pios_com_telem_rf_id;
+			break;
+		case HWSHARED_PORTTYPES_GPS:
+			PIOS_HAL_configure_com(usart_port_cfg, PIOS_COM_GPS_RX_BUF_LEN, PIOS_COM_GPS_TX_BUF_LEN, com_driver, &port_driver_id);
+			target = &pios_com_gps_id;
+			break;
+		case HWSHARED_PORTTYPES_DSM:
+#if defined(PIOS_INCLUDE_DSM)
+			if (dsm_cfg && usart_dsm_hsum_cfg) {
+				PIOS_HAL_configure_dsm(usart_dsm_hsum_cfg, dsm_cfg, com_driver, dsm_bind);
+			}
+#endif  /* PIOS_INCLUDE_DSM */
+			break;
+		case HWSHARED_PORTTYPES_SBUS:
+#if defined(PIOS_INCLUDE_SBUS) && defined(PIOS_INCLUDE_USART)
+			if (sbus_cfg && sbus_rcvr_cfg) {
+				uintptr_t usart_sbus_id;
+				if (PIOS_USART_Init(&usart_sbus_id, sbus_rcvr_cfg)) {
+					PIOS_Assert(0);
+				}
+				uintptr_t sbus_id;
+				if (PIOS_SBus_Init(&sbus_id, sbus_cfg, com_driver, usart_sbus_id)) {
```

Comment:
Right now the child port type would not include SBUS in those cases.  If it erroneously did, we would be prevented from configuring it by the if block.

I don't know about soft usart, but I don't know why sbus can't be "bring your own inverter" like the frsky modes etc.  Just takes documentation.  But that's for another PR.